[JSC] Add NoIndexing miss optimization

Constellation opened the pull request:
#### 801dcc1c9e7aabec321390689bc45e35752f371b
<pre>
[JSC] Add NoIndexing miss optimization
<a href="https://bugs.webkit.org/show_bug.cgi?id=240290">https://bugs.webkit.org/show_bug.cgi?id=240290</a>
rdar://problem/93456330

Reviewed by Saam Barati.

`Array.from({ length: 42 })` idiom becomes common. And this is currently inefficient since we
perform missed indexed access of the object 42 times. But if we do not have indexed properties
for this object and it is plain normal object, then we can have optimization folding it to undefined
as we are skipping prototype-chain access for OOB array access.

This patch introduces IndexedNoIndexingMiss IC. It guards via structures, and they are ensured that
they do not have indexing types. To represent it in PropertyCondition, this patch addds AbsenceOfIndexedProperties
PropertyCondition. When we found NoIndexing, we attempt to create IC by collecting structures.

In the future, we would like to further optimize it in DFG and FTL by using this condition. But for now,
we just use IC. We also add simple optimization via AI-proven structures.

PropertyCondition AbsenceOfIndexedProperties is useful, and in the future patches, we should use it to ensure
that iterator protocol is met more efficiently.

Attached benchmark gets 7x performance improvement.

                                   ToT                     Patched

array-from-object           697.6093+-0.9029     ^     89.7427+-0.7996        ^ definitely 7.7734x faster
array-from-object-func      698.9900+-0.5077     ^    103.7614+-1.7910        ^ definitely 6.7365x faster

* JSTests/microbenchmarks/array-from-derived-object-func.js: Added.
(shouldBe):
(Derived):
* JSTests/microbenchmarks/array-from-object-func.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-from-object.js: Added.
(shouldBe):
* JSTests/stress/no-indexing-shape-invalidate.js: Added.
(shouldBe):
(test):
* JSTests/stress/no-indexing-shape-multiple.js: Added.
(shouldBe):
(test):
* JSTests/stress/no-indexing-shape-negative.js: Added.
(shouldBe):
(test):
* JSTests/stress/no-indexing-shape-other-array.js: Added.
(shouldBe):
(test):
* JSTests/stress/no-indexing-shape-prototype-invalidate.js: Added.
(shouldBe):
(test):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::add8):
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::generateWithGuard):
(JSC::AccessCase::generateImpl):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
(JSC::AccessCase::create): Deleted.
* Source/JavaScriptCore/bytecode/ArrayProfile.h:
(JSC::ArrayProfile::ArrayProfile): Deleted.
* Source/JavaScriptCore/bytecode/ObjectPropertyCondition.h:
(JSC::ObjectPropertyCondition::absenceOfIndexedPropertiesWithoutBarrier):
(JSC::ObjectPropertyCondition::absenceOfIndexedProperties):
* Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.cpp:
(JSC::generateConditionsForIndexedMiss):
* Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.h:
* Source/JavaScriptCore/bytecode/PolymorphicAccess.cpp:
(JSC::PolymorphicAccess::regenerate):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/PropertyCondition.cpp:
(JSC::PropertyCondition::dumpInContext const):
(JSC::PropertyCondition::isStillValidAssumingImpurePropertyWatchpoint const):
(JSC::PropertyCondition::validityRequiresImpurePropertyWatchpoint const):
(JSC::PropertyCondition::isStillValid const):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/PropertyCondition.h:
(JSC::PropertyCondition::absenceOfIndexedPropertiesWithoutBarrier):
(JSC::PropertyCondition::absenceOfIndexedProperties):
(JSC::PropertyCondition::hasPrototype const):
(JSC::PropertyCondition::hash const):
(JSC::PropertyCondition::operator== const):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheArrayGetByVal):

Canonical link: <a href="https://commits.webkit.org/253120@main">https://commits.webkit.org/253120@main</a>
</pre>

saambarati commented on Source/JavaScriptCore/bytecode/ObjectPropertyCondition.h:
```diff
@@ -120,7 +120,22 @@ class ObjectPropertyCondition {
     }
-    
+
+    static ObjectPropertyCondition absenceOfIndexedWithoutBarrier(JSObject* object, JSObject* prototype)
```
name nit: Can we call this absenceOfIndexedProperties in all the various function/enum names?

Constellation replied:
Changed.

saambarati commented on Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.cpp:
```diff
@@ -491,6 +515,20 @@ ObjectPropertyConditionSet generateConditionsForPropertySetterMissConcurrently(
+            return true;
+        });
+}
+
```
no need for this function since it seems like it's not used.

Constellation replied:
Yeah, this is not used right now.

saambarati commented on Source/JavaScriptCore/bytecode/PropertyCondition.cpp:
```diff
@@ -224,7 +226,46 @@ bool PropertyCondition::isStillValidAssumingImpurePropertyWatchpoint(
+            if (PropertyConditionInternal::verbose)
+                dataLog("Invalid because it's a dictionary.\n");
+            return false;
+        }
```
We still transition a dictionary if we add an indexing type, right?

Constellation replied:
Ah, right, we can remove it.

saambarati commented on Source/JavaScriptCore/bytecode/PropertyCondition.cpp:
```diff
@@ -337,6 +379,7 @@ bool PropertyCondition::isStillValid(Concurrency concurrency, Structure* structu
     // absence. It doesn't affect AbsenceOfSetEffect because impure properties aren't ever setters.
     switch (m_header.type()) {
     case Absence:
+    case AbsenceOfIndexed:
```
is this needed for indexed accesses?

saambarati replied:
If this is just you being conservative, maybe just say this might be more conservative than we need to be

Constellation replied:
Yeah, right now it is added conservatively. I'll drop it.

saambarati commented on Source/JavaScriptCore/bytecode/Repatch.cpp:
```diff
@@ -665,6 +665,38 @@ static InlineCacheAction tryCacheArrayGetByVal(JSGlobalObject* globalObject, Cod
+                if (!base->isObject())
+                    return GiveUpOnCache;
+
+                if (base->structure()->mayInterceptIndexedAccesses() || base->structure()->typeInfo().interceptsGetOwnPropertySlotByIndexEvenWhenLengthIsNotZero())
+                    return GiveUpOnCache;
```
won't this be handled in the `generateConditionsForIndexedMiss` call?

Constellation replied:
`generateConditions...` functions only checks Structures of prototype or upper. So for instance's Structure, we need to check it here before calling it.

saambarati replied:
right, makes sense

saambarati commented on Source/JavaScriptCore/bytecode/Repatch.cpp:
```diff
@@ -665,6 +665,38 @@ static InlineCacheAction tryCacheArrayGetByVal(JSGlobalObject* globalObject, Cod
+                if (base->structure()->mayInterceptIndexedAccesses() || base->structure()->typeInfo().interceptsGetOwnPropertySlotByIndexEvenWhenLengthIsNotZero())
+                    return GiveUpOnCache;
+
+                auto cacheStatus = prepareChainForCaching(globalObject, base, nullptr);
```
Do we need prepareChainForCaching for indexed accesses? Some things in the loop seem like they might be needed, but others don't. For example, I don't think we need to flatten dictionaries, right? Same with looking at `propertyAccessesAreCacheable` I think. But we might need the isProxy check. I guess it's ok to be conservative and do this for now, even if it's more conservative than it needs to be.

Constellation replied:
For now, I think this is good, getting `usesPolyProto` status etc. It is a bit conservative. I'll put FIXME for the future improvement.

saambarati replied:
sounds good

saambarati commented on Source/JavaScriptCore/bytecode/Repatch.cpp:
```diff
@@ -665,6 +665,38 @@ static InlineCacheAction tryCacheArrayGetByVal(JSGlobalObject* globalObject, Cod
+                    return GiveUpOnCache;
+
+                auto newCase = AccessCase::create(vm, codeBlock, AccessCase::IndexedNoIndexingMiss, nullptr, invalidOffset, headStructure, conditionSet);
+                result = stubInfo.addAccessCase(locker, globalObject, codeBlock, ECMAMode::strict(), nullptr, WTFMove(newCase));
```
How do we know it's strict? This seems like something the access site should pass in

Constellation replied:
This `strict` or not `strict` matters only for `put` operation, so here, it does not matter.
It is aligned to the existing code in L705.

saambarati commented on Source/JavaScriptCore/bytecode/Repatch.cpp:
```diff
@@ -665,6 +665,38 @@ static InlineCacheAction tryCacheArrayGetByVal(JSGlobalObject* globalObject, Cod
+                }
+
+                fireWatchpointsAndClearStubIfNeeded(vm, stubInfo, codeBlock, result);
+                return result.shouldGiveUpNow() ? GiveUpOnCache : RetryCacheLater;
```
this is a lot of duplicated code for no reason. I suggest adopting the idiom we do in other repatch code where we have a unique_ptr on AccessCase, and if it's null, we call create below. If it's non-null, we don't call create. That way we don't have this copy pasted code here.

Constellation replied:
That sounds good. will change.

saambarati commented on Source/JavaScriptCore/bytecode/AccessCase.cpp:
```diff
@@ -1322,6 +1330,13 @@ void AccessCase::generateWithGuard(
         return;
     }
 
+    case IndexedNoIndexingMiss: {
+        emitDefaultGuard();
+        GPRReg propertyGPR = state.propertyGPR();
+        state.failAndIgnore.append(jit.branch32(CCallHelpers::LessThan, propertyGPR, CCallHelpers::TrustedImm32(0)));
+        break;
+    }
```
nice

saambarati commented on Source/JavaScriptCore/bytecode/AccessCase.cpp:
```diff
@@ -2493,7 +2508,12 @@ void AccessCase::generateImpl(AccessGenerationState& state)
     }
-        
+
+    case IndexedNoIndexingMiss:
+        jit.moveTrustedValue(jsUndefined(), valueRegs);
+        state.succeed();
+        return;
```
nice

saambarati reviewed (approved):
r=me with your fixes to the comments

Constellation committed:
[JSC] Add NoIndexing miss optimization
https://bugs.webkit.org/show_bug.cgi?id=240290
rdar://problem/93456330

Reviewed by Saam Barati.

`Array.from({ length: 42 })` idiom becomes common. And this is currently inefficient since we
perform missed indexed access of the object 42 times. But if we do not have indexed properties
for this object and it is plain normal object, then we can have optimization folding it to undefined
as we are skipping prototype-chain access for OOB array access.

This patch introduces IndexedNoIndexingMiss IC. It guards via structures, and they are ensured that
they do not have indexing types. To represent it in PropertyCondition, this patch addds AbsenceOfIndexedProperties
PropertyCondition. When we found NoIndexing, we attempt to create IC by collecting structures.

In the future, we would like to further optimize it in DFG and FTL by using this condition. But for now,
we just use IC. We also add simple optimization via AI-proven structures.

PropertyCondition AbsenceOfIndexedProperties is useful, and in the future patches, we should use it to ensure
that iterator protocol is met more efficiently.

Attached benchmark gets 7x performance improvement.

                                   ToT                     Patched

array-from-object           697.6093+-0.9029     ^     89.7427+-0.7996        ^ definitely 7.7734x faster
array-from-object-func      698.9900+-0.5077     ^    103.7614+-1.7910        ^ definitely 6.7365x faster

* JSTests/microbenchmarks/array-from-derived-object-func.js: Added.
(shouldBe):
(Derived):
* JSTests/microbenchmarks/array-from-object-func.js: Added.
(shouldBe):
* JSTests/microbenchmarks/array-from-object.js: Added.
(shouldBe):
* JSTests/stress/no-indexing-shape-invalidate.js: Added.
(shouldBe):
(test):
* JSTests/stress/no-indexing-shape-multiple.js: Added.
(shouldBe):
(test):
* JSTests/stress/no-indexing-shape-negative.js: Added.
(shouldBe):
(test):
* JSTests/stress/no-indexing-shape-other-array.js: Added.
(shouldBe):
(test):
* JSTests/stress/no-indexing-shape-prototype-invalidate.js: Added.
(shouldBe):
(test):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::add8):
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::generateWithGuard):
(JSC::AccessCase::generateImpl):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
(JSC::AccessCase::create): Deleted.
* Source/JavaScriptCore/bytecode/ArrayProfile.h:
(JSC::ArrayProfile::ArrayProfile): Deleted.
* Source/JavaScriptCore/bytecode/ObjectPropertyCondition.h:
(JSC::ObjectPropertyCondition::absenceOfIndexedPropertiesWithoutBarrier):
(JSC::ObjectPropertyCondition::absenceOfIndexedProperties):
* Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.cpp:
(JSC::generateConditionsForIndexedMiss):
* Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.h:
* Source/JavaScriptCore/bytecode/PolymorphicAccess.cpp:
(JSC::PolymorphicAccess::regenerate):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/PropertyCondition.cpp:
(JSC::PropertyCondition::dumpInContext const):
(JSC::PropertyCondition::isStillValidAssumingImpurePropertyWatchpoint const):
(JSC::PropertyCondition::validityRequiresImpurePropertyWatchpoint const):
(JSC::PropertyCondition::isStillValid const):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/PropertyCondition.h:
(JSC::PropertyCondition::absenceOfIndexedPropertiesWithoutBarrier):
(JSC::PropertyCondition::absenceOfIndexedProperties):
(JSC::PropertyCondition::hasPrototype const):
(JSC::PropertyCondition::hash const):
(JSC::PropertyCondition::operator== const):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheArrayGetByVal):

Canonical link: https://commits.webkit.org/253120@main
